pkg/networking/backend_sg_provider.go

```diff
@@ -0,0 +1,245 @@
+	if !p.backendSGEnabled {
+		return nil, nil
+	}
+	if len(p.backendSGs) > 0 {
+		return p.backendSGs, nil
+	}
+	// Auto generate Backend Security group, and return the id
+	if err := p.allocateBackendSG(ctx); err != nil {
+		p.logger.Error(err, "Failed to auto-create backend SG")
+		return nil, err
+	}
+	return []string{p.autoGeneratedSG}, nil
+}
+
+func (p *defaultBackendSGProvider) Release(ctx context.Context) error {
+	if !p.backendSGEnabled || len(p.backendSGs) > 0 {
+		return nil
+	}
+	if required, err := p.isBackendSGRequired(ctx); required || err != nil {
+		return err
+	}
+	return p.releaseSG(ctx)
+}
+
+func (p *defaultBackendSGProvider) allocateBackendSG(ctx context.Context) error {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	if len(p.autoGeneratedSG) > 0 {
+		return nil
+	}
+
+	sgName := p.getBackendSGName()
+	sgID, err := p.getBackendSGFromEC2(ctx, sgName, p.vpcID)
+	if err != nil {
+		return err
+	}
+	if len(sgID) > 1 {
+		p.logger.V(1).Info("Existing SG found", "id", sgID)
+		p.autoGeneratedSG = sgID
+		return nil
+	}
+
+	createReq := &ec2sdk.CreateSecurityGroupInput{
+		VpcId:             awssdk.String(p.vpcID),
+		GroupName:         awssdk.String(sgName),
+		Description:       awssdk.String(sgDescription),
+		TagSpecifications: p.buildBackendSGTags(ctx),
+	}
+	p.logger.Info("creating securityGroup", "name", sgName)
+	resp, err := p.ec2Client.CreateSecurityGroupWithContext(ctx, createReq)
+	if err != nil {
+		return err
+	}
+	p.logger.V(1).Info("created SecurityGroup", "name", sgName, "id", resp.GroupId)
+	p.autoGeneratedSG = awssdk.StringValue(resp.GroupId)
+	return nil
+}
+
+func (p *defaultBackendSGProvider) buildBackendSGTags(_ context.Context) []*ec2sdk.TagSpecification {
+	return []*ec2sdk.TagSpecification{
+		{
+			ResourceType: awssdk.String(resourceTypeSecurityGroup),
+			Tags: []*ec2sdk.Tag{
+				{
+					Key:   awssdk.String(tagKeyK8sCluster),
+					Value: awssdk.String(p.clusterName),
+				},
+				{
+					Key:   awssdk.String(tagKeySGType),
+					Value: awssdk.String(tagValueBackend),
+				},
+			},
+		},
+	}
+}
+
+func (p *defaultBackendSGProvider) getBackendSGFromEC2(ctx context.Context, sgName string, vpcID string) (string, error) {
+	req := &ec2sdk.DescribeSecurityGroupsInput{
+		Filters: []*ec2sdk.Filter{
+			{
+				Name:   awssdk.String("vpc-id"),
+				Values: awssdk.StringSlice([]string{vpcID}),
+			},
+			{
+				Name:   awssdk.String("group-name"),
+				Values: awssdk.StringSlice([]string{sgName}),
+			},
+		},
+	}
+	p.logger.V(1).Info("Queriying existing SG", "vpc-id", vpcID, "name", sgName)
+	sgs, err := p.ec2Client.DescribeSecurityGroupsAsList(ctx, req)
+	if err != nil && !isEC2SecurityGroupNotFoundError(err) {
+		return "", err
+	}
+	if len(sgs) > 0 {
+		return awssdk.StringValue(sgs[0].GroupId), nil
+	}
+	return "", nil
+}
+
+func (p *defaultBackendSGProvider) isBackendSGRequired(ctx context.Context) (bool, error) {
+	ingList := &networking.IngressList{}
```

Comment:
personally i feel this depends on ingress implementation details too much.

also, seems the backendSG is not cleaned if there are some ingress in cluster even the left Ingresses don't need the backendSG.

shall we have a dedicated finalizer so that only ingresses/services needs backSG is added these finalizers?
if we have that, we can have a dedicated function to test whether ingress/service needs backendSG, and then add this dedicated finalizer and invoke backendSGProvider.Get inside controller, and pass obtained backendSG into modelBuilder.
and then call backendSGProvider.Release if any group member have this finalizer, and remove the finalizer if it succeeds.